# ENGR 1451/2451 — Weekly Assignment 1
**10 points**

This assignment focuses on four goals: basic Python expressions, working with a `pandas` DataFrame, calculating summary statistics, and visualizing data.

**Due:** See Canvas.

Complete the work directly in this notebook. Show your Python code and give brief written explanations where requested. Submit the completed notebook as directed in Canvas.

## What to hand in

Work this assignment with the **course tutor** in the class project. That is the AI help to use on this assignment, and it keeps the record for you — you do not write one yourself.

At the end of each working session, type **"make my record"** in the chat. Copy the block it produces and paste it into the **Session records** cell near the bottom of this notebook.

You do not have to finish in one sitting. To pick up later, start a new chat in the same project, say which assignment you are on and which problem you stopped at, and keep going. That session gives you its own block, which you paste under the previous one.

Records are not graded for correctness or for how much help you needed, and extra credit is given for a record that shows you attempted the problem. The code and the written answers still have to be yours.

## LE1-1. Python warm-up (1 pt)
Use Python as a calculator. In one code cell, show an example of:
- addition,
- subtraction,
- multiplication,
- division, and
- exponentiation.

In [1]:
add = 1+1
subtract = 1-2
mult = 1*4
divis = 6/3
expon = 2 ** 4
print("1 + 1 =", add, 
      "| 1 - 2 =", subtract, 
      "| 1 * 4 =", mult,
      "| 6/3 =", divis,
     "| 2^4 =", expon)

1 + 1 = 2 | 1 - 2 = -1 | 1 * 4 = 4 | 6/3 = 2.0 | 2^4 = 16


## LE1-2. Iris data: structure, summary statistics, and visualization (6 pts)

A useful data-analysis workflow can be summarized as:
**question → data → cleaning → exploratory analysis → visualization → interpretation → communication → reproducible code**.

For this exercise, use the `iris.csv` file supplied with this assignment. Download it from Canvas and put it in a folder named `data` next to this notebook, so that the path `./data/iris.csv` works. The file contains measurements of sepal length/width and petal length/width for three iris species.

In [6]:
import pandas as pd
import matplotlib.pyplot as plt

iris = pd.read_csv("./data/iris.csv")
iris = iris.drop(columns=["rownames"], errors="ignore")
iris.head()

,Sepal.Length,Sepal.Width,Petal.Length,Petal.Width,Species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa
3,4.6,3.1,1.5,0.2,setosa
4,5.0,3.6,1.4,0.2,setosa


### A. Understand the data (2 pts)
Using Python, answer:
1. How many observations (rows) are in the dataset?
2. Which variables are numerical? Are they continuous or discrete?
3. Which variable is categorical, and what are its categories?
4. How many observations belong to each species?

**Answer:**

In [27]:
# Useful tools: iris.shape, iris.dtypes, iris.select_dtypes(),
# iris["Species"].unique(), iris["Species"].value_counts()

# Your code here
#Loading in data
import pandas as pd
iris = pd.read_csv("./data/iris.csv")

#1
print("#1")
rows,cols = iris.shape
print("There are", rows, "rows in the dataset.")

#2
print("\n#2")
numeric_col_names = iris.select_dtypes(include="number").columns.tolist() #all numeric vars
continuous_vars_col_names = iris.select_dtypes(float).columns.tolist() #continuous vars (unless if whole numbers are reported as 3.0000--i.e. with unneded decimal places)
discrete_vars_col_names = iris.select_dtypes(int).columns.tolist() #discrete vars

listvars = [continuous_vars_col_names, discrete_vars_col_names] #creating a list of lists to index through 
for i in range(len(listvars)):
    if len(listvars[i]) == 0: #checking for empty lists
        listvar = "none"

print("The following variables are numerical: ", ", ".join(numeric_col_names),
      "\nContinuous variables:", ", ".join(continuous_vars_col_names), "\nDiscrete variables:", ", ".join(discrete_vars_col_names))

#3
print("\n#3")
categorical_col_names = iris.select_dtypes(include="string")

#4
print("\n#4")


#1
There are 150 rows in the dataset.

#2
The following variables are numerical:  rownames, Sepal.Length, Sepal.Width, Petal.Length, Petal.Width 
Continuous variables: Sepal.Length, Sepal.Width, Petal.Length, Petal.Width 
Discrete variables: rownames

#3

#4


### B. Analyze *versicolor* (2 pts)
1. Create a DataFrame containing only `versicolor` flowers.
2. Report the mean and median of each numerical variable.
3. Briefly explain why a single overall mean for all three species can be misleading.

**Answer:**

In [ ]:
# Your code here
#1
print("#1")

#2
print("\n#2")

#3
print("\n#3")

### C. Visualize the iris data (2 pts)
Create a scatter plot of **Petal.Length** versus **Petal.Width** for all flowers in the iris dataset.

1. Put `Petal.Width` on the x-axis and `Petal.Length` on the y-axis.
2. Color the points by species so the three species can be compared.
3. Label both axes and include a legend.
4. Briefly describe the main pattern you see in the plot.

**Answer:**

In [28]:
# Your visualization code here
#1
print("#1")

#2
print("\n#2")

#3
print("\n#3")

#4
print("\n#4")

## LE1-3. Distribution shape and appropriate statistics (3 pts)

For each situation, state:
1. expected shape: **symmetric, right-skewed, or left-skewed**;
2. better measure of center: **mean or median**;
3. better measure of spread: **standard deviation or IQR**;
4. a brief reason.

Each case is worth 0.75 pt.

| Case | Situation |
|---|---|
| **a** | House prices: Q1 = \$350k, median = \$450k, Q3 = \$1.0M, with a meaningful number above \$6M |
| **b** | House prices: Q1 = \$300k, median = \$600k, Q3 = \$900k, with very few above \$1.2M |
| **c** | Weekly alcoholic drinks among students: most consume 0, while a few consume many |
| **d** | Scores on a straightforward 100-point quiz: most students score in the 80s and 90s, with a few much lower scores |

# **Answers:**

(for 4., reasoning is included in parts 1-3)
## a. 
    1. Right-skewed, larger delta between median and Q3 (as opposed to Q1 and median) + meaningful number of high price (>6M) outliers indicates that there will be a significant skew towards the right/upper end of house prices.
    2. Given the prescence of high outliers and noticeable skew, the median will be a better measure of central tendency than the mean. The median is more resistant to outliers than the mean.
    3. Just like the median, the IQR is more resistant to outliers and skew than the standard deviation, so it will likley be the better measure of spread in this case.

## b.
    1. Symmetric, the delta between the median and Q1/Q2 is identical and there are very few high outliers, which points towards a symmetric shape.
    2. Since the data is symmetric and has few outliers, the mean would likley be the best measure of central tendency in this case. 
    3. Just like the mean, since the data is symmetric and has few outliers, the standard deviation would likley be the best measure of central tendency in this case.
    4. To add onto 2. & 3., the median and IQR are based on rank, whereas the mean and standard deviation take into account every data point. In this case, the data is symmetric and has few outliers, so the mean and standard deviation will provide a slightly more representative measure of central tendency and spread than the median and IQR. 
    
## c.
    1. Right-skewed, most of the data (weekly alcoholic drinks among students) is concentrated around 0, while a few outliers consume a larger amount of alcoholic drinks.
    2. (same explanation as 2 & 3 from part a.) Given the prescence of outliers and noticeable righ-skew, the median will be a better measure of central tendency than the mean. The median is more resistant to outliers than the mean.
    3. (same explanation as 2 & 3 from part a.)Just like the median, the IQR is more resistant to outliers and skew than the standard deviation, so it will likley be the better measure of spread in this case.
    
## d. 
    1. Left-skewed, most of the data is concentrated around higher scores, but a few outliers score very low.
    2. (same explanation as 2 & 3 from part a.) Given the prescence of outliers and noticeable left-skew, the median will be a better measure of central tendency than the mean. The median is more resistant to outliers than the mean.
    3. (same explanation as 2 & 3 from part a.) Just like the median, the IQR is more resistant to outliers and skew than the standard deviation, so it will likley be the better measure of spread in this case.

In [3]:
## Session records

At the end of each working session, type **"make my record"** in the chat and paste the block it produces into the cell below. If you work across several sessions, paste each new block underneath the previous one.

Assignments are not graded for correctness or for how much help you needed, and extra credit is given for a record that shows you attempted the problem. They are read to find out which steps are hardest, and used to improve the course.

SyntaxError: invalid syntax (3284297802.py, line 3)

{
  "schema_version": "engr-tutor-export-2",
  "assignment": "LE01",
  "export_n": 1,
  "session_date": "2026-08-29",
  "messages": [
    {"i": 1, "who": "student", "text": "Hello! I am starting LEO1."},
    {"i": 2, "who": "tutor", "served": ["how_to_use", "a1_preamble", "a1_le1_1"]},
    {"i": 3, "who": "student", "text": "Is exp() the exponentiation command in python?"},
    {"i": 4, "who": "tutor", "text": "exp() is not the exponentiation operator... math.exp(x) computes e**x. The operator for exponentiation in Python is **, e.g. 2 ** 3 gives 8. Note ^ is bitwise XOR in Python, not exponentiation. [trimmed] Want me to say more about ** with negative or fractional exponents, or are you set to write your cell?"},
    {"i": 5, "who": "student", "text": "Understood"},
    {"i": 6, "who": "tutor", "text": "Good — sounds like you're ready. Go ahead and write your cell with all five operators whenever you want, and paste it here if you'd like a look, or just move on to LE1-2 when you're set."},
    {"i": 7, "who": "student", "text": "How can I express strings and numeric variables with the print command?"},
    {"i": 8, "who": "tutor", "text": "print() takes one or more things separated by commas... [trimmed examples of printing strings, numbers, mixing text and numbers with commas, and labeling results like print(\"Sum:\", a + b)] ...also f-strings, e.g. print(f\"Sum: {a + b}\"). Want me to go further into f-strings, or are you good to write your LE1-1 cell now?"},
    {"i": 9, "who": "student", "text": "Is ... the command for continuing a long string?"},
    {"i": 10, "who": "tutor", "text": "... is not how Python continues a string across lines; it's the Ellipsis object, unrelated to strings. [trimmed — explained backslash line continuation, implicit continuation inside parentheses, and adjacent string literal joining, with examples] For something as short as your LE1-1 line you probably won't need any of this. Want me to go further, or are you set?"},
    {"i": 11, "who": "student", "text": "Got it"},
    {"i": 12, "who": "tutor", "text": "Sounds good — go ahead and write your LE1-1 cell whenever you're ready. Paste it here if you want a look, or let me know when you're moving on to LE1-2."},
    {"i": 13, "who": "student", "text": "Does one code cell mean including every desired operation in the markdown cell or everything in one line of code?"},
    {"i": 14, "who": "tutor", "text": "\"One code cell\" refers to the code cell, not the markdown cell above it. Within that one cell you don't need everything on one line — five separate print() lines, one per operation, is the natural way. [trimmed example] Does that clear it up, or is something still fuzzy?"},
    {"i": 15, "who": "student", "text": "How can I install pandas on jupyter? I have it installed on my computer, but do I need to install it in the terminal on the Unix teaching cluster?"},
    {"i": 16, "who": "tutor", "text": "Having pandas locally doesn't help if the kernel runs on the cluster — that's a separate machine. [trimmed — check with import pandas as pd; print(pd.__version__), then pip install pandas --user on a cluster terminal if missing, then confirm the notebook's kernel matches the environment where it was installed] If you hit a specific error, paste it here."},
    {"i": 17, "who": "student", "text": "I accidentally ran pip install panda, how can I delete the panda library from the teaching cluster?"},
    {"i": 18, "who": "tutor", "text": "That installed an unrelated package named panda, not pandas. Run pip uninstall panda (or pip uninstall panda -y to skip confirmation) in a cluster terminal session. Won't affect pandas at all. Let me know if the uninstall throws anything unexpected."},
    {"i": 19, "who": "student", "text": "Cool, that worked"},
    {"i": 20, "who": "tutor", "text": "Great, glad that's sorted. Go ahead and get pandas installed and imported whenever you're ready, and let me know how LE1-1 or LE1-2 is going."},
    {"i": 21, "who": "student", "text": "Will (row,col) = iris.shape save the row and column numbers in the row and col variables?"},
    {"i": 22, "who": "tutor", "text": "Yes — .shape returns a tuple like (150, 5), and Python supports tuple unpacking, so row, col = iris.shape assigns row=150, col=5 by position. Parentheses are optional. Number of variables must match the tuple length or you get a ValueError. [trimmed generic example] Does that make sense, or do you want to see what happens with a longer tuple?"},
    {"i": 23, "who": "student", "text": "Nice, can you run over what this code is doing and how I could recreate it in future data tasks? \niris = iris.drop(columns=[\"rownames\"], errors=\"ignore\")\niris.head()", "code": true},
    {"i": 24, "who": "tutor", "text": ".drop() removes rows or columns; columns= targets column(s), passed as a list even for one column. It returns a new DataFrame rather than modifying in place, which is why the line reassigns iris. errors=\"ignore\" prevents a KeyError if the column isn't there. .head() returns the first 5 rows by default, a quick sanity check. [trimmed generic example on a made-up df] This drop(columns=[...], errors=\"ignore\") pattern is a solid default for safely dropping a column that might already be gone. Want me to go over .tail() or .sample() too?"},
    {"i": 25, "who": "student", "text": "Can you help visualize this more for me?"},
    {"i": 26, "who": "tutor", "text": "The left box is df before the drop — three columns, three rows. The id column gets dropped, and what comes out the other side is a new DataFrame with only name and score. The original df reference is reassigned to point at that new, smaller frame — nothing is modified in place. Then .head() on the result just returns its first rows for a quick look. Same shape you'd use for iris, just with rownames in place of id. Want to look at anything else before moving on?"},
    {"i": 27, "who": "student", "text": "Can you explain the difference between discrete and continuous variables?"},
    {"i": 28, "who": "tutor", "text": "Discrete variables take separate, countable values with gaps between them — number of siblings, petal count. Continuous variables can take any value in a range, limited only by measurement precision — height, weight, time. Test: if there's an infinite number of possible values between two nearby measurements, it's continuous. A wrinkle: a column stored with a fixed number of decimals doesn't by itself tell you discrete vs continuous — that's about what's being measured, not how it's rounded for display. Does that distinction feel solid, or want to try classifying a few examples together first?"},
    {"i": 29, "who": "student", "text": "That makes sense"},
    {"i": 30, "who": "tutor", "text": "Good — sounds like you've got the concept down. Whenever you're ready, go ahead and apply it to iris's numerical columns for LE1-2A, and let me know how it goes or if anything else comes up."},
    {"i": 31, "who": "student", "text": "Which command can I use to identify numeric variables? I am thinking I could use iris.select_dtypes() to manually select int, float, and double, but I was wondering if there was a more efficient way of achieving the same goal."},
    {"i": 32, "who": "tutor", "text": "select_dtypes() takes include=, and you can pass the general category \"number\" instead of listing every specific numeric type: df.select_dtypes(include=\"number\"). [trimmed generic example] include=\"object\" or exclude=\"number\" gets the complement; .columns on the result gives just the names. Want to see exclude= too, or are you set to use this on iris?"},
    {"i": 33, "who": "student", "text": "What does include= do in python?"},
    {"i": 34, "who": "tutor", "text": "include= isn't a general Python keyword, it's a parameter name defined by select_dtypes() specifically. For select_dtypes() it tells the method which column types to keep — a string like \"number\", a list of types, or an actual type like int. [trimmed generic example with a greet(name, greeting=\"Hello\") function showing named parameters] Does that clear it up, or do you want to look at exclude= too?"},
    {"i": 35, "who": "student", "text": "Is there a command I can use to extract the titles of the rows of the numeric variables and store them as a list of string variables?"},
    {"i": 36, "who": "tutor", "text": "df.select_dtypes(include=\"number\").columns.tolist() — .columns gives a pandas Index, .tolist() converts it to a plain Python list. [trimmed generic example] Note: \"titles of the rows\" isn't quite right terminology — these are column names, not row titles; rows are usually just numbered by index. Does that distinction make sense, or want to see how row indices work too?"},
    {"i": 37, "who": "student", "text": "Could I use cont_vars_col_names = iris.select_dtypes(float).columns.tolist to identifiy continuous variables in the dataset?", "code": true},
    {"i": 38, "who": "tutor", "text": "select_dtypes(float) works fine positionally, but .tolist is missing its parentheses — without () you're referencing the method itself, not calling it. Fix: iris.select_dtypes(float).columns.tolist(). Separately: dtype and continuous-vs-discrete aren't strictly the same thing in general — a discrete variable could be stored as float, and vice versa — worth checking dtype against what's actually being measured for iris specifically. Want to check iris.dtypes side by side with what each column measures before committing to that list?"},
    {"i": 39, "who": "student", "text": "Will python return an error if it does not find any float variables in the dataset?"},
    {"i": 40, "who": "tutor", "text": "No — select_dtypes(float) with no matches just returns an empty DataFrame, and .columns.tolist() on that gives an empty list, []. No crash. [trimmed generic example] Worth running that same line on iris to see what actually comes back. Want to talk through what to do if the list comes back empty when you didn't expect it to?"},
    {"i": 41, "who": "student", "text": "Can I use \nif(listvar) == empty\n    listvar = \"no variables\"\nTo set listvar to \"no variables\" if the list is empty", "code": true},
    {"i": 42, "who": "tutor", "text": "Good instinct on the logic, but a few syntax issues: empty isn't a Python keyword — check length or rely on empty lists being falsy. Every if needs a colon and an indented body. == is comparison, = is assignment (you had that right on the second line). Fix: if len(listvar) == 0: listvar = \"no variables\" — or the more idiomatic if not listvar: listvar = \"no variables\". [trimmed generic example] Does that clear up the syntax, or want to see how this behaves for a list like [0] or [\"\"]?"},
    {"i": 43, "who": "student", "text": "Got it"},
    {"i": 44, "who": "tutor", "text": "Sounds good. Go ahead and try that on your iris numeric-columns list, and let me know if you hit anything else or want to move on to the rest of LE1-2A."},
    {"i": 45, "who": "student", "text": "Can I use listvars = [cont_vars_col_names, disc_vars_col_names] to make a list of lists?", "code": true},
    {"i": 46, "who": "tutor", "text": "Yes — putting two existing lists inside [], separated by a comma, gives a list of lists. listvars[0] gets the continuous list, listvars[1] gets the discrete list. [trimmed generic example] Worth flagging: a dictionary like {\"continuous\": cont_vars_col_names, \"discrete\": disc_vars_col_names} might read more clearly than remembering index 0 vs 1, but not required. Want to see more of dictionaries, or are you set with the list-of-lists version?"},
    {"i": 47, "who": "student", "text": "Will this code work for incrementing through listvars and checking for empty lists?\nlistvars = [cont_vars_col_names, disc_vars_col_names] #creating a list of lists to index through \nfor i in len(listvars):\n    if len(listvar) == 0: #checking for empty lists\n        listvar = \"no variables\"", "code": true},
    {"i": 48, "who": "tutor", "text": "Close in intent, two issues: len(listvars) is a single number, not iterable — you want range(len(listvars)) to generate index positions 0, 1. Also you're checking listvar (no s), which was never defined; since you're looping by index i, you want listvars[i] for both the check and the reassignment. [trimmed generic example on data = [[1,2,3], [], [\"a\",\"b\"]]] Also, for item in data: is more direct when you don't need i, but since you're reassigning in place, indexing is the right tool here. Want to try fixing your version with these two changes, or see the indexing-vs-direct-iteration distinction laid out more?"},
    {"i": 49, "who": "student", "text": "Cool thanks"},
    {"i": 50, "who": "tutor", "text": "Sounds good — go ahead and give it a try with those two fixes, and paste it here if you want a check, or let me know how it runs."},
    {"i": 51, "who": "student", "text": "Does /n work in python print commands?"},
    {"i": 52, "who": "tutor", "text": "No — /n is just the literal characters slash and n. The actual newline is \\n (backslash), an escape sequence: print(\"Line one\\nLine two\") breaks onto two lines. Other common escapes: \\t (tab), \\\\ (literal backslash), \\\" (literal quote). [trimmed generic comparison example] Want to see other escape sequences, or are you set?"},
    {"i": 53, "who": "student", "text": "How can I get lists to display nicely in print commands?"},
    {"i": 54, "who": "tutor", "text": "A few options: default print(nums) is plain but works; \", \".join(...) gives a clean separator without brackets (numbers need str() conversion first); looping and printing one item per line; combining with an f-string label; pprint from the pprint module for nested structures like lists of lists. [trimmed examples of each] For listvars specifically, .join() on each inner list, or labeled print() calls (\"Continuous:\", \"Discrete:\"), is probably cleanest. Want to see how to label each inner list as you print it?"},
    {"i": 55, "who": "student", "text": "make my record"}
  ],
  "events": [
    {"i": 4, "part": "1-1", "type": "explain"},
    {"i": 8, "part": "1-1", "type": "explain"},
    {"i": 10, "part": "1-1", "type": "explain"},
    {"i": 14, "part": "1-1", "type": "explain"},
    {"i": 16, "part": "1-2A", "type": "explain"},
    {"i": 18, "part": "1-2A", "type": "explain"},
    {"i": 22, "part": "1-2A", "type": "explain"},
    {"i": 23, "part": "1-2A", "type": "code"},
    {"i": 24, "part": "1-2A", "type": "explain"},
    {"i": 26, "part": "1-2A", "type": "explain"},
    {"i": 28, "part": "1-2A", "type": "explain"},
    {"i": 32, "part": "1-2A", "type": "explain"},
    {"i": 34, "part": "1-2A", "type": "explain"},
    {"i": 36, "part": "1-2A", "type": "explain"},
    {"i": 37, "part": "1-2A", "type": "code"},
    {"i": 38, "part": "1-2A", "type": "explain"},
    {"i": 40, "part": "1-2A", "type": "explain"},
    {"i": 41, "part": "1-2A", "type": "code"},
    {"i": 42, "part": "1-2A", "type": "explain"},
    {"i": 45, "part": "1-2A", "type": "code"},
    {"i": 46, "part": "1-2A", "type": "explain"},
    {"i": 47, "part": "1-2A", "type": "code"},
    {"i": 48, "part": "1-2A", "type": "explain"},
    {"i": 52, "part": "1-2A", "type": "explain"},
    {"i": 54, "part": "1-2A", "type": "explain"}
  ],
  "parts": [
    {"part": "1-1", "exchanges": 6, "max_tier": 1, "passed": false},
    {"part": "1-2A", "exchanges": 17, "max_tier": 1, "passed": false}
  ],
  "tutor_assessment": [
    "The longest stretch was building the toolchain for LE1-2A — select_dtypes, .columns.tolist(), an empty-list check, a list of lists, and finally a for-loop over indices; the recurring sticking point was small syntax slips (missing parentheses on .tolist, len() where range(len()) was needed, a name mismatch between listvar and listvars[i]) rather than confusion about the underlying pandas or Python concepts.",
    "No release valve or 'I got it, let's move on' was invoked this session, so there's no signal of unhelpfulness to flag; the student never reached stated answers for LE1-2A's four questions or wrote any explanatory prose, so that work is still ahead of them next session."
  ]
}
<!-- Paste your session record blocks in this cell, one after another. -->
{
  "schema_version": "engr-tutor-export-2",
  "assignment": "LE01",
  "export_n": 2,
  "session_date": "2026-08-31",
  "messages": [
    {"i": 1, "who": "student", "text": "Hello! I am starting LEO1."},
    {"i": 2, "who": "tutor", "served": ["how_to_use", "a1_preamble", "a1_le1_1"]},
    {"i": 3, "who": "student", "text": "Is exp() the exponentiation command in python?"},
    {"i": 4, "who": "tutor", "text": "exp() is not the exponentiation operator... math.exp(x) computes e**x. The operator for exponentiation in Python is **, e.g. 2 ** 3 gives 8. Note ^ is bitwise XOR in Python, not exponentiation. [trimmed] Want me to say more about ** with negative or fractional exponents, or are you set to write your cell?"},
    {"i": 5, "who": "student", "text": "Understood"},
    {"i": 6, "who": "tutor", "text": "Good — sounds like you're ready. Go ahead and write your cell with all five operators whenever you want, and paste it here if you'd like a look, or just move on to LE1-2 when you're set."},
    {"i": 7, "who": "student", "text": "How can I express strings and numeric variables with the print command?"},
    {"i": 8, "who": "tutor", "text": "print() takes one or more things separated by commas... [trimmed examples of printing strings, numbers, mixing text and numbers with commas, and labeling results like print(\"Sum:\", a + b)] ...also f-strings, e.g. print(f\"Sum: {a + b}\"). Want me to go further into f-strings, or are you good to write your LE1-1 cell now?"},
    {"i": 9, "who": "student", "text": "Is ... the command for continuing a long string?"},
    {"i": 10, "who": "tutor", "text": "... is not how Python continues a string across lines; it's the Ellipsis object, unrelated to strings. [trimmed — explained backslash line continuation, implicit continuation inside parentheses, and adjacent string literal joining, with examples] For something as short as your LE1-1 line you probably won't need any of this. Want me to go further, or are you set?"},
    {"i": 11, "who": "student", "text": "Got it"},
    {"i": 12, "who": "tutor", "text": "Sounds good — go ahead and write your LE1-1 cell whenever you're ready. Paste it here if you want a look, or let me know when you're moving on to LE1-2."},
    {"i": 13, "who": "student", "text": "Does one code cell mean including every desired operation in the markdown cell or everything in one line of code?"},
    {"i": 14, "who": "tutor", "text": "\"One code cell\" refers to the code cell, not the markdown cell above it. Within that one cell you don't need everything on one line — five separate print() lines, one per operation, is the natural way. [trimmed example] Does that clear it up, or is something still fuzzy?"},
    {"i": 15, "who": "student", "text": "How can I install pandas on jupyter? I have it installed on my computer, but do I need to install it in the terminal on the Unix teaching cluster?"},
    {"i": 16, "who": "tutor", "text": "Having pandas locally doesn't help if the kernel runs on the cluster — that's a separate machine. [trimmed — check with import pandas as pd; print(pd.__version__), then pip install pandas --user on a cluster terminal if missing, then confirm the notebook's kernel matches the environment where it was installed] If you hit a specific error, paste it here."},
    {"i": 17, "who": "student", "text": "I accidentally ran pip install panda, how can I delete the panda library from the teaching cluster?"},
    {"i": 18, "who": "tutor", "text": "That installed an unrelated package named panda, not pandas. Run pip uninstall panda (or pip uninstall panda -y to skip confirmation) in a cluster terminal session. Won't affect pandas at all. Let me know if the uninstall throws anything unexpected."},
    {"i": 19, "who": "student", "text": "Cool, that worked"},
    {"i": 20, "who": "tutor", "text": "Great, glad that's sorted. Go ahead and get pandas installed and imported whenever you're ready, and let me know how LE1-1 or LE1-2 is going."},
    {"i": 21, "who": "student", "text": "Will (row,col) = iris.shape save the row and column numbers in the row and col variables?"},
    {"i": 22, "who": "tutor", "text": "Yes — .shape returns a tuple like (150, 5), and Python supports tuple unpacking, so row, col = iris.shape assigns row=150, col=5 by position. Parentheses are optional. Number of variables must match the tuple length or you get a ValueError. [trimmed generic example] Does that make sense, or do you want to see what happens with a longer tuple?"},
    {"i": 23, "who": "student", "text": "Nice, can you run over what this code is doing and how I could recreate it in future data tasks? \niris = iris.drop(columns=[\"rownames\"], errors=\"ignore\")\niris.head()", "code": true},
    {"i": 24, "who": "tutor", "text": ".drop() removes rows or columns; columns= targets column(s), passed as a list even for one column. It returns a new DataFrame rather than modifying in place, which is why the line reassigns iris. errors=\"ignore\" prevents a KeyError if the column isn't there. .head() returns the first 5 rows by default, a quick sanity check. [trimmed generic example on a made-up df] This drop(columns=[...], errors=\"ignore\") pattern is a solid default for safely dropping a column that might already be gone. Want me to go over .tail() or .sample() too?"},
    {"i": 25, "who": "student", "text": "Can you help visualize this more for me?"},
    {"i": 26, "who": "tutor", "text": "The left box is df before the drop — three columns, three rows. The id column gets dropped, and what comes out the other side is a new DataFrame with only name and score. The original df reference is reassigned to point at that new, smaller frame — nothing is modified in place. Then .head() on the result just returns its first rows for a quick look. Same shape you'd use for iris, just with rownames in place of id. Want to look at anything else before moving on?"},
    {"i": 27, "who": "student", "text": "Can you explain the difference between discrete and continuous variables?"},
    {"i": 28, "who": "tutor", "text": "Discrete variables take separate, countable values with gaps between them — number of siblings, petal count. Continuous variables can take any value in a range, limited only by measurement precision — height, weight, time. Test: if there's an infinite number of possible values between two nearby measurements, it's continuous. A wrinkle: a column stored with a fixed number of decimals doesn't by itself tell you discrete vs continuous — that's about what's being measured, not how it's rounded for display. Does that distinction feel solid, or want to try classifying a few examples together first?"},
    {"i": 29, "who": "student", "text": "That makes sense"},
    {"i": 30, "who": "tutor", "text": "Good — sounds like you've got the concept down. Whenever you're ready, go ahead and apply it to iris's numerical columns for LE1-2A, and let me know how it goes or if anything else comes up."},
    {"i": 31, "who": "student", "text": "Which command can I use to identify numeric variables? I am thinking I could use iris.select_dtypes() to manually select int, float, and double, but I was wondering if there was a more efficient way of achieving the same goal."},
    {"i": 32, "who": "tutor", "text": "select_dtypes() takes include=, and you can pass the general category \"number\" instead of listing every specific numeric type: df.select_dtypes(include=\"number\"). [trimmed generic example] include=\"object\" or exclude=\"number\" gets the complement; .columns on the result gives just the names. Want to see exclude= too, or are you set to use this on iris?"},
    {"i": 33, "who": "student", "text": "What does include= do in python?"},
    {"i": 34, "who": "tutor", "text": "include= isn't a general Python keyword, it's a parameter name defined by select_dtypes() specifically. For select_dtypes() it tells the method which column types to keep — a string like \"number\", a list of types, or an actual type like int. [trimmed generic example with a greet(name, greeting=\"Hello\") function showing named parameters] Does that clear it up, or do you want to look at exclude= too?"},
    {"i": 35, "who": "student", "text": "Is there a command I can use to extract the titles of the rows of the numeric variables and store them as a list of string variables?"},
    {"i": 36, "who": "tutor", "text": "df.select_dtypes(include=\"number\").columns.tolist() — .columns gives a pandas Index, .tolist() converts it to a plain Python list. [trimmed generic example] Note: \"titles of the rows\" isn't quite right terminology — these are column names, not row titles; rows are usually just numbered by index. Does that distinction make sense, or want to see how row indices work too?"},
    {"i": 37, "who": "student", "text": "Could I use cont_vars_col_names = iris.select_dtypes(float).columns.tolist to identifiy continuous variables in the dataset?", "code": true},
    {"i": 38, "who": "tutor", "text": "select_dtypes(float) works fine positionally, but .tolist is missing its parentheses — without () you're referencing the method itself, not calling it. Fix: iris.select_dtypes(float).columns.tolist(). Separately: dtype and continuous-vs-discrete aren't strictly the same thing in general — a discrete variable could be stored as float, and vice versa — worth checking dtype against what's actually being measured for iris specifically. Want to check iris.dtypes side by side with what each column measures before committing to that list?"},
    {"i": 39, "who": "student", "text": "Will python return an error if it does not find any float variables in the dataset?"},
    {"i": 40, "who": "tutor", "text": "No — select_dtypes(float) with no matches just returns an empty DataFrame, and .columns.tolist() on that gives an empty list, []. No crash. [trimmed generic example] Worth running that same line on iris to see what actually comes back. Want to talk through what to do if the list comes back empty when you didn't expect it to?"},
    {"i": 41, "who": "student", "text": "Can I use \nif(listvar) == empty\n    listvar = \"no variables\"\nTo set listvar to \"no variables\" if the list is empty", "code": true},
    {"i": 42, "who": "tutor", "text": "Good instinct on the logic, but a few syntax issues: empty isn't a Python keyword — check length or rely on empty lists being falsy. Every if needs a colon and an indented body. == is comparison, = is assignment (you had that right on the second line). Fix: if len(listvar) == 0: listvar = \"no variables\" — or the more idiomatic if not listvar: listvar = \"no variables\". [trimmed generic example] Does that clear up the syntax, or want to see how this behaves for a list like [0] or [\"\"]?"},
    {"i": 43, "who": "student", "text": "Got it"},
    {"i": 44, "who": "tutor", "text": "Sounds good. Go ahead and try that on your iris numeric-columns list, and let me know if you hit anything else or want to move on to the rest of LE1-2A."},
    {"i": 45, "who": "student", "text": "Can I use listvars = [cont_vars_col_names, disc_vars_col_names] to make a list of lists?", "code": true},
    {"i": 46, "who": "tutor", "text": "Yes — putting two existing lists inside [], separated by a comma, gives a list of lists. listvars[0] gets the continuous list, listvars[1] gets the discrete list. [trimmed generic example] Worth flagging: a dictionary like {\"continuous\": cont_vars_col_names, \"discrete\": disc_vars_col_names} might read more clearly than remembering index 0 vs 1, but not required. Want to see more of dictionaries, or are you set with the list-of-lists version?"},
    {"i": 47, "who": "student", "text": "Will this code work for incrementing through listvars and checking for empty lists?\nlistvars = [cont_vars_col_names, disc_vars_col_names] #creating a list of lists to index through \nfor i in len(listvars):\n    if len(listvar) == 0: #checking for empty lists\n        listvar = \"no variables\"", "code": true},
    {"i": 48, "who": "tutor", "text": "Close in intent, two issues: len(listvars) is a single number, not iterable — you want range(len(listvars)) to generate index positions 0, 1. Also you're checking listvar (no s), which was never defined; since you're looping by index i, you want listvars[i] for both the check and the reassignment. [trimmed generic example on data = [[1,2,3], [], [\"a\",\"b\"]]] Also, for item in data: is more direct when you don't need i, but since you're reassigning in place, indexing is the right tool here. Want to try fixing your version with these two changes, or see the indexing-vs-direct-iteration distinction laid out more?"},
    {"i": 49, "who": "student", "text": "Cool thanks"},
    {"i": 50, "who": "tutor", "text": "Sounds good — go ahead and give it a try with those two fixes, and paste it here if you want a check, or let me know how it runs."},
    {"i": 51, "who": "student", "text": "Does /n work in python print commands?"},
    {"i": 52, "who": "tutor", "text": "No — /n is just the literal characters slash and n. The actual newline is \\n (backslash), an escape sequence: print(\"Line one\\nLine two\") breaks onto two lines. Other common escapes: \\t (tab), \\\\ (literal backslash), \\\" (literal quote). [trimmed generic comparison example] Want to see other escape sequences, or are you set?"},
    {"i": 53, "who": "student", "text": "How can I get lists to display nicely in print commands?"},
    {"i": 54, "who": "tutor", "text": "A few options: default print(nums) is plain but works; \", \".join(...) gives a clean separator without brackets (numbers need str() conversion first); looping and printing one item per line; combining with an f-string label; pprint from the pprint module for nested structures like lists of lists. [trimmed examples of each] For listvars specifically, .join() on each inner list, or labeled print() calls (\"Continuous:\", \"Discrete:\"), is probably cleanest. Want to see how to label each inner list as you print it?"},
    {"i": 55, "who": "student", "text": "make my record"},
    {"i": 56, "who": "tutor", "text": "[export_n: 1 emitted]"},
    {"i": 57, "who": "student", "text": "can you illustrate the difference between left and right skew for me?"},
    {"i": 58, "who": "tutor", "text": "[rendered a diagram widget: left-skewed vs right-skewed histograms with mean/median markers] The naming convention trips people up: the skew direction refers to which side the long tail points, not where most of the data sits. In the left-skewed panel, most values cluster on the right, but a few low outliers stretch the tail out to the left — and that tail pulls the mean down below the median. Right-skewed is the mirror image. That pull on the mean is exactly why the choice between mean and median matters for a skewed distribution — useful to have in mind for LE1-3. Want to look at anything else, or are you set to work through those cases yourself?"},
    {"i": 59, "who": "student", "text": "Can you make the skew in the graphs more pronounced?"},
    {"i": 60, "who": "tutor", "text": "[rendered a more exaggerated version of the same diagram] That's more exaggerated now — a tight cluster of tall bars with a thin scattering of small bars trailing far off to one side. The bigger the gap between mean and median lines, the more pronounced the skew. Let me know if you want to work through anything else, or if you're ready to apply this to LE1-3."},
    {"i": 61, "who": "student", "text": "Can you create visual graphs for part LE1-3 to help assist with visualization?"},
    {"i": 62, "who": "tutor", "text": "That's where I need to hold back a bit — drawing the actual distribution shape for each of the four LE1-3 cases would be handing you the answer to the first thing each case asks you to determine. What I can do is show general shapes so you know what each looks like, then it's on you to picture each scenario and decide which shape fits. [trimmed] Want that, or do you want to take a first pass at the four cases yourself?"},
    {"i": 63, "who": "student", "text": "How can I create a text cell in Jupyter?"},
    {"i": 64, "who": "tutor", "text": "Cells are code cells by default; press m in command mode to convert to Markdown, or use the toolbar dropdown. b/a to create a new cell below/above first. Shift+Enter to run/render. [trimmed] For LE1-3 specifically, you've already got Markdown cells set up under each Answer heading — click into those directly. Let me know if you run into anything else."},
    {"i": 65, "who": "student", "text": "Is this an appropriate analysis of part 1. of a.?"},
    {"i": 66, "who": "tutor", "text": "Happy to look — but I don't see your analysis pasted in yet. Go ahead and paste what you wrote for part 1 of case (a) and I'll take a look."},
    {"i": 67, "who": "student", "text": "Right-skewed, larger delta between median and Q3 (as opposed to Q1 and median) + large number of high price (>6M) outliers indicates that the average house price will likley be greater than the median, resulting in a right-skewed data set"},
    {"i": 68, "who": "tutor", "text": "That's a real analysis, not just a guess — you're using both pieces of evidence: the quartile spacing and the outliers. [trimmed] The correct relationship for right-skewed: a long right tail pulls the mean up relative to the median. One thing worth double-checking: you framed the conclusion as mean greater than median causing right-skew, but the shape is really the cause and the mean-median relationship is the effect. Does the causal direction there matter to you, or is that just a wording thing? Separate from that: you've nailed part 1. Parts 2, 3, 4 still need your answer. Want to work through those next?"},
    {"i": 69, "who": "student", "text": "Thank you, I always mix up left and right skew"},
    {"i": 70, "who": "tutor", "text": "That mix-up is one of the most common ones in stats. The trick: skew is named for the direction the tail points, not where most of the data piles up — a right-skewed distribution has its bulk on the left and a thin tail stretching right, toward the high outliers. For case (a), that thin tail toward the $6M+ prices is what makes it right-skewed, which lines up with what you already worked out. Want to move on to parts 2-4 for case (a), or take a look at case (b) first since the two are meant to be compared?"},
    {"i": 71, "who": "student", "text": "I refined my explanation to focus more on shape: Right-skewed, larger delta between median and Q3 (as opposed to Q1 and median) + large number of high price (>6M) outliers indicates that there will be a significant skew towards the right/upper end of house prices."},
    {"i": 72, "who": "tutor", "text": "That fixes exactly the thing we talked about — now the shape is the thing your evidence points to directly, rather than routing through the mean-median comparison first. Good revision. Parts 2 and 3 for case (a) — which measure of center and which measure of spread fit a right-skewed distribution with outliers like this — plus part 4, the reason. Want to take a pass at those, or talk through how to think about center and spread first?"},
    {"i": 73, "who": "student", "text": "Given the prescence of high outliers and noticeable skew, the median will be a better measure of central tendency than the mean. The median is more resistant to outliers than the mean."},
    {"i": 74, "who": "tutor", "text": "That's solid reasoning and correctly stated — the median resists being pulled by outliers the way the mean does, so median is the right call here. Parts 2 and 3 are actually two separate questions though — you've answered part 2 (center: median) but part 3 asks about spread (standard deviation or IQR). Want to take a pass at that one too, or think out loud about how outliers affect a spread measure the same way they affect a measure of center?"},
    {"i": 75, "who": "student", "text": "Just like the median, the IQR is more resistant to outliers and skew than the standard deviation, so it will likley be the better measure of spread in this case."},
    {"i": 76, "who": "tutor", "text": "Nice — that's the same resistance-to-outliers logic carried over correctly, and it's the right measure. That's all four parts of case (a): right-skewed, median, IQR, and your reasons for each. Case (b) is next if you want to keep going — it's a mirror comparison, so worth noticing what's different about its numbers this time. Want to tackle it, or take a break here?"},
    {"i": 77, "who": "student", "text": "I am going to do the rest now and come back if I need to check my work, thanks!"},
    {"i": 78, "who": "tutor", "text": "Sounds good — go ahead and work through the rest at your own pace. Paste anything here if you want a check along the way, or when you're done for the session (or done with LE1-3 entirely), just say \"make my record\" and I'll put together the block for your notebook."},
    {"i": 79, "who": "student", "text": "Is the mean ever a better measure of central tendency than the median? In larger symmetric datasets, do the mean and median not generally fall around the same value?"},
    {"i": 80, "who": "tutor", "text": "Yes to both. For symmetric distributions without significant outliers, the mean is generally preferred since it uses every data point's actual value, not just rank; the median only cares about the middle position and discards distance information. [trimmed] In a perfectly symmetric distribution, mean and median are theoretically equal; real data rarely gets there exactly, but the more symmetric and larger the dataset, the closer they sit. So: skew or outliers -> median; roughly symmetric and outlier-free -> mean. That's useful background for cases (c) and (d), where you'll picture the shape from a written description rather than a five-number summary."},
    {"i": 81, "who": "student", "text": "Are there other measures of central tendency that combine rank and all values in the dataset?"},
    {"i": 82, "who": "tutor", "text": "Yes — trimmed (truncated) mean removes a fixed percentage of extreme values from both ends before averaging the rest. Winsorized mean caps extreme values at a percentile rather than removing them, so they still contribute. Weighted mean uses all values but gives some more influence via assigned weights (not really about outlier resistance). Mode is the most frequent value, ignoring both rank and magnitude — mentioned for completeness though not directly relevant here. [trimmed] For this course, mean and median are the two you're asked to choose between in LE1-3. Want to see a quick numeric example of how a trimmed mean handles an outlier differently than a plain mean?"},
    {"i": 83, "who": "student", "text": "All good, thank you! Just curious"},
    {"i": 84, "who": "tutor", "text": "Glad it scratched the itch. Go ahead and keep working through the rest of LE1-3 — ping me if you want anything checked, or when you're ready to wrap up for the session."},
    {"i": 85, "who": "student", "text": "make my record"}
  ],
  "events": [
    {"i": 4, "part": "1-1", "type": "explain"},
    {"i": 8, "part": "1-1", "type": "explain"},
    {"i": 10, "part": "1-1", "type": "explain"},
    {"i": 14, "part": "1-1", "type": "explain"},
    {"i": 16, "part": "1-2A", "type": "explain"},
    {"i": 18, "part": "1-2A", "type": "explain"},
    {"i": 22, "part": "1-2A", "type": "explain"},
    {"i": 23, "part": "1-2A", "type": "code"},
    {"i": 24, "part": "1-2A", "type": "explain"},
    {"i": 26, "part": "1-2A", "type": "explain"},
    {"i": 28, "part": "1-2A", "type": "explain"},
    {"i": 32, "part": "1-2A", "type": "explain"},
    {"i": 34, "part": "1-2A", "type": "explain"},
    {"i": 36, "part": "1-2A", "type": "explain"},
    {"i": 37, "part": "1-2A", "type": "code"},
    {"i": 38, "part": "1-2A", "type": "explain"},
    {"i": 40, "part": "1-2A", "type": "explain"},
    {"i": 41, "part": "1-2A", "type": "code"},
    {"i": 42, "part": "1-2A", "type": "explain"},
    {"i": 45, "part": "1-2A", "type": "code"},
    {"i": 46, "part": "1-2A", "type": "explain"},
    {"i": 47, "part": "1-2A", "type": "code"},
    {"i": 48, "part": "1-2A", "type": "explain"},
    {"i": 52, "part": "1-2A", "type": "explain"},
    {"i": 54, "part": "1-2A", "type": "explain"},
    {"i": 67, "part": "1-3", "type": "code"},
    {"i": 68, "part": "1-3", "type": "part_pass"},
    {"i": 71, "part": "1-3", "type": "code"},
    {"i": 73, "part": "1-3", "type": "code"},
    {"i": 75, "part": "1-3", "type": "code"},
    {"i": 76, "part": "1-3", "type": "part_pass"},
    {"i": 80, "part": "1-3", "type": "explain"},
    {"i": 82, "part": "1-3", "type": "explain"}
  ],
  "parts": [
    {"part": "1-1", "exchanges": 6, "max_tier": 1, "passed": false},
    {"part": "1-2A", "exchanges": 17, "max_tier": 1, "passed": false},
    {"part": "1-3", "exchanges": 6, "max_tier": 2, "passed": true}
  ],
  "tutor_assessment": [
    "The longest stretch remains LE1-2A's toolchain (select_dtypes, .tolist(), empty-list checks, list of lists, index-based looping), sticking on small syntax slips rather than conceptual confusion; case (a) of LE1-3 went smoothly once the causal-direction wording was tightened, with the student independently revising their own answer after one nudge and then correctly reasoning through center and spread with only light confirmation needed.",
    "No release valve or 'I got it, let's move on' was invoked this session, so there is no signal of tutor unhelpfulness to flag; LE1-1 and the rest of LE1-2A (parts 2-4) and LE1-3 cases (b), (c), (d) remain unfinished going into the next session."
  ]
}
*(session records go here)*

## Before you submit

1. **Restart Kernel and Run All Cells** and confirm every cell ran without errors.
2. Confirm your plot and printed values are visible in the saved file.
3. Confirm your session record blocks are pasted into the cell above.
4. Save the notebook, then upload the `.ipynb` to Canvas.

### Reference
Adapted from exercises in *OpenIntro Statistics*, 4th ed., and the original course assignment.